### PR TITLE
Add CSS Standards & Guidelines for 3.3

### PR DESCRIPTION
**Pull request by hanakin**
This includes all the coding standards we check for via stylelint. as well as all the documentation on the coding structure and practices we will be using for the new front-end

### CODE REVIEW

File: development/development/css_guidelines.rst [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        width: 100px;
+        height: 100px;
+        margin: 0;
+        padding: 8px;
+
+        /* Misc */
+        content: "-";
+    }
+
+Proper Use of units
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: `Proper use of units`

File: development/development/css_guidelines.rst [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+~~~~~~~~~~~~~~~~~~~
+
+CSS allows for the use of several different unit types. As such it can get confusing when using more than one type of unit through out the project. For that reason its beneficial to stick to a stick set of rules for what unit types are to be used for certain selectors.
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: `throughout`, `it's`

File: development/development/css_guidelines.rst [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+
+Furthermore there are certain reasons to use or avoid using specific units in certain places.
+
+EM
+^^
+
+The ‘em’ unit. This is a very problematic unit which reeks havoc on countless projects due to the way its calculated. As such this unit type must be avoid except for very very minimal use cases. We prevent the use of ``em`` except for ``letter-spaceing`` & ``word-spacing``. It is also used for icon sizing but that is an edge case.
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: `reeks`? I think it should be`wreaks`

Comment (marc1706) [reply development/development/css_guidelines.rst]: Also, `letter-spacing`

File: development/development/css_guidelines.rst [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+
+Line-heights
+^^^^^^^^^^^^
+
+All line-heights are to be specified as ``unitless`` in order to prevent in proper inheritance. By nature when using units with line-heights the children inherit by default. This can lead to unwanted effects and bloated code. A ``sass`` function called ``unitless`` is provided which will convert px values for convenience, but for clarity the math is simply
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: `in proper` --> `improper`

File: development/development/css_guidelines.rst [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+
+Font-size
+^^^^^^^^^
+
+All ``font-size`` should be specified either in ``px`` or ``%`` in small cases. All px values will be converted to ``rem`` during the build process as ``rem`` provide for control in responsive situations.
+
+Margins & Paddings
+^^^^^^^^^^^^^^^^^^
+
+All ``margin`` & ``padding`` should be specified in ``px`` values or ``%``. All ``px`` All px values will be converted to ``rem`` during the build process as ``rem`` provide for control in responsive situations.
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: `All px` is written twice

File: development/development/css_guidelines.rst [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+
+PX
+^^
+
+All ``px`` will be whole numbers. Browsers do not render ``px`` in fractional values despite what you browser may say it is. Only calculated values will display as fractional ``px``. For clarification a calculated value would be units like ``rem``, ``em``, ``%``, & even ``unitless`` as is the case with line-heights.
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: what you --> what your

File: development/development/css_guidelines.rst [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+
+Dimensions
+^^^^^^^^^^
+
+All dimensional values ``width``, ``min-width``, ``height``, & ``min-height`` should be specified in ``px`` or ``%``. A case can be made for ``vw`` & ``vh``, but they are still on the fringe of browser acceptance, as such fallbacks in ``px`` or ``%`` are required. These values will remain as px if specified. This is done as ``height`` is more effectively and appropriately controlled via the ``line-height`` property, and ``width`` is better specified using the objects box-model via ``padding`` unless its fluid in which ``100%`` can be specified or u can also use ``left: 0; right: 0;``
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: or u --> or you

File: development/development/css_guidelines.rst [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+    .foo {
+        color: red;
+
+        .bar {
+            color: blue;
+        }
+    }
+
+…we will be left with this compiled CSS:
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: space before we

File: development/development/css_guidelines.rst [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+.. code:: css
+
+    .foo { color: red; }
+    .foo .bar { color: blue; }
+
+When indenting Sass, we stick to the same two indentation, and we also leave a blank line before and after the nested ruleset.
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: two indentations?

File: development/development/css_guidelines.rst [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Our project makes use of several tools to lint and to keep us to the standards.
+
+1. `stylelint.io`_
+^^^^^^^^^^^^^^^^^^
+
+.. note::
+
+This is used to provide detailed linting for our standards via the ``.stlyelintrc`` file in the root of the project.
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: `.stlyelintrc` --> `.stylelintrc`

File: development/development/css_guidelines.rst [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+.. note::
+
+This is used to provide automatic sorting to our declaration order via the ``.postcss-sorting.json`` file in the root of the project.
+
+3. `postcss-pxtorem`_
+^^^^^^^^^^^^^^^^^^^^^
+
+.. note::
+
+This is used to ensure the proper units are consistently used throughout the project during the build process via the ``gulp`` as well as on save in your editor.
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: via `gulp`

File: development/development/css_guidelines.rst [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+4. `stylefmt`_
+^^^^^^^^^^^^^^
+
+.. note::
+
+This is used to help automatically re-format your code to the standards on the fly during the build process via ``gulp`` as well as on save in your editor.
+
+.. note::
+
+As a **NOTE** our editor of choice is `ATOM`_ which provides usefull plugins to make use of these tools. Checkout the `Editor Setup`_ section of the docs for more information
```
Comment (marc1706) [inline development/development/css_guidelines.rst]: We also use PhpStorm but oh well

File: development/development/css_naming.rst [hunk 13]
```diff
@@ -0,0 +1,168 @@
+Naming Conventions
+------------------
+
+Naming conventions in CSS are hugely useful in making your code more strict, more transparent, and more informative.
+
+A good naming convention will tell you and your team
+
+-  what type of thing a class does;
```
Comment (marc1706) [inline development/development/css_naming.rst]: remove semi-colons

File: development/development/css_naming.rst [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+-  where a class can be used;
+-  what (else) a class might be related to.
+
+The naming convention we follow are as follows
+
+-  Hyphen (``-``) delimited strings.
+-  Layer namespacing
+-  A variation on BEM-like naming for action modifiers
+
+It’s worth noting that not all the naming convention are normally useful in the CSS-side of development; they really come into their own when viewed in HTML.
```
Comment (marc1706) [inline development/development/css_naming.rst]: all the naming conventions

File: development/development/css_naming.rst [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+A good naming convention will tell you and your team
+
+-  what type of thing a class does;
+-  where a class can be used;
+-  what (else) a class might be related to.
+
+The naming convention we follow are as follows
+
+-  Hyphen (``-``) delimited strings.
```
Comment (marc1706) [inline development/development/css_naming.rst]: remove full stop

File: development/development/css_naming.rst [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+'''''''''''''''
+
+   -  `UI Selector Namspacing`_
+
+JavaScript Hooks
+~~~~~~~~~~~~~~~~
+
+As a rule, it is unwise to bind your CSS and your JS onto the same class in your HTML. This is because doing so means you can’t have (or remove) one without (removing) the other. It is much cleaner, much more transparent, and much more maintainable to bind your JS onto specific classes.
+
+I have known occasions before when trying to refactor some CSS has unwittingly removed JS functionality because the two were tied to each other—it was impossible to have one without the other.
```
Comment (marc1706) [inline development/development/css_naming.rst]: Kind of odd to read and not sure if it adds any info.

File: development/development/css_naming.rst [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+
+Typically, these are classes that are prepended with ``js-``, for
+
+example:
```
Comment (marc1706) [inline development/development/css_naming.rst]: shouldn't be on a new line

File: development/development/css_naming.rst [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+
+.. code:: html
+
+    <input type="submit" class="btn js-btn" value="Follow" />
+
+This means that we can have an element elsewhere which can carry with style of ``.btn {}``, but without the behavior of ``.js-btn``.
```
Comment (marc1706) [inline development/development/css_naming.rst]: carry the style of `.btn {}`

File: development/development/css_guidelines.rst [hunk 19]
```diff
@@ -0,0 +1,292 @@
+Introduction
+------------
+
+CSS is not a pretty language. While it is simple to learn and get started with, it soon becomes problematic at any reasonable scale. There isn’t much we can do to change how CSS works, but we can make changes to the way we author and structure it.
+
+There are a variety of techniques we must employ in order to satisfy these goals, and CSS Guidelines is a document of recommendations and approaches that will help us to do so.
+
+The first part of this will deal with syntax, formatting and CSS anatomy, the second part will deal with approach, mindframe and attitude toward writing and architecting CSS.
```
[the remark this answers is not in the document]
Comment (hanakin) [reply development/development/css_guidelines.rst]: I did not write it all myself that's why there are different styles.

File: development/development/css_naming.rst (hunk 16 above)
Comment (hanakin) [reply development/development/css_naming.rst]: this whole section is wrong anyway so I will clean it up in the re-write

File: development/development/css_selectors.rst [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+
+    /**
+     * If the element is a `div`, also give it some box-like styling.
+     */
+    div.error {
+        padding: 10px;
+        border: 1px solid;
+    }
+
+This is one example where a qualified selector might be justifiable, but I would still recommend an approach more like:
```
Comment (marc1706) [inline development/development/css_selectors.rst]: Try not to use I/we: `, but an approach like the following is still recommended:`

Comment (marc1706) [reply development/development/css_selectors.rst]: Although I have to admit that there is probably a lot more that could be changed to get rid of I/we

File: development/development/css_selectors.rst [hunk 21]
```diff
@@ -0,0 +1,256 @@
...
+
+Naming
+~~~~~~
+
+As Phil Karlton once said
+
+    ‘There are only two hard things in Computer Science: cache
+    invalidation and naming things.’
+
+I won’t comment on the former claim here, but the latter has plagued me for years. My advice with regard to naming things in CSS is to pick a name that is sensible, but somewhat ambiguous: aim for high reusability. For example, instead of a class like ``.site-nav``, choose something like ``.primary-nav``; rather than ``.footer-links``, favor a class like ``.sub-links``.
```
Comment (marc1706) [inline development/development/css_selectors.rst]: Same here, probably better to get rid of I as this documentation is seen as the phpBB documentation and not by a specific person.

File: development/development/css_selectors.rst [hunk 22]
```diff
@@ -0,0 +1,256 @@
...
+
+    Tying your class name semantics tightly to the nature of the content
+    has already reduced the ability of your architecture to scale or be
+    easily put to use by other developers.
+
+That is to say, we should use sensible names—classes like ``.border`` or ``.red`` are never advisable—but we should avoid using classes which describe the exact nature of the content and/or its use cases. **Using a class name to describe content is redundant because content describes itself.**
+
+The debate surrounding semantics has raged for years, but it is important that we adopt a more pragmatic, sensible approach to naming things in order to work more efficiently and effectively. Instead of focussing on ‘semantics’,look more closely at sensibility and longevity—choose names based on ease of maintenance, not for their perceived meaning.
+
+Name things for people; they’re the only things that actually read your classes (everything else merely matches them). Once again, it is better to strive for reusable, recyclable classes rather than writing for specific use cases. Let’s take an example:
```
Comment (marc1706) [inline development/development/css_selectors.rst]: `only ones that actually read`

**Comment by hanakin** [comment]
@marc1706 css_selectors had a lot of both mine and Harry Roberts so there was a lot of switching focus, I opted more for the user you/your and scrubbed everything

**Comment by hanakin** [comment]
@marc1706 when you get time can you finish the rest of this review?

**Comment by marc1706** [comment]
I've found some time to read through the rest and I think it's good to go.

**Pull request merged**